Mathlib/Data/Polynomial/Degree/Definitions.lean

```diff
@@ -0,0 +1,1582 @@
+/-
+Copyright (c) 2018 Chris Hughes. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Hughes, Johannes Hölzl, Scott Morrison, Jens Wagemaker
+
+! This file was ported from Lean 3 source module data.polynomial.degree.definitions
+! leanprover-community/mathlib commit 808ea4ebfabeb599f21ec4ae87d6dc969597887f
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.Data.Fintype.BigOperators
+import Mathlib.Data.Nat.WithBot
+import Mathlib.Data.Polynomial.Monomial
+import Mathlib.Data.Polynomial.Coeff
+
+/-!
+# Theory of univariate polynomials
+
+The definitions include
+`degree`, `Monic`, `leadingCoeff`
+
+Results include
+- `degree_mul` : The degree of the product is the sum of degrees
+- `leadingCoeff_add_of_degree_eq` and `leadingCoeff_add_of_degree_lt` :
+    The leading_coefficient of a sum is determined by the leading coefficients and degrees
+-/
+
+
+set_option linter.uppercaseLean3 false
+
+noncomputable section
+
+open Finsupp Finset
+
+open BigOperators Classical Polynomial
+
+namespace Polynomial
+
+universe u v
+
+variable {R : Type u} {S : Type v} {a b c d : R} {n m : ℕ}
+
+section Semiring
+
+variable [Semiring R] {p q r : R[X]}
+
+-- Porting note: In Lean3, these are same `coe`, but not in Lean4.
+private theorem _root_.WithBot.some_eq_nat_cast : WithBot.some n = Nat.cast n := rfl
```

Comment:
This already exists as `Nat.cast_withBot` in `Mathlib/Data/Nat/Cast/WithTop.lean`